### PR TITLE
Url detection

### PR DESCRIPTION
**Pull request by michaelcfanning**
## Changes

Add a simple URL analysis that's currently constrained to a small subset of code files. We could consider expanding this and adding a dynamic validator (that can do things like provide a broken link detection).

@suvamM

### CODE REVIEW

File: Src/Plugins/Security/SEC102.ReviewPotentiallySensitiveData.json [hunk 1]
```diff
@@ -22,6 +24,13 @@
           "Name": "ReviewPotentiallySensitiveData/SocialSecurityNumber",
           "ContentsRegex": "\\b(?P<refine>(?P<secret>[0-9]{3}-[0-9]{2}-[0-9]{4}))\\b",
           "MessageArguments": { "dataKind": "social security number" }
+        },
+        {
+          "Id": "SEC102/003",
```
Comment (michaelcfanning) [inline Src/Plugins/Security/SEC102.ReviewPotentiallySensitiveData.json:29]: >Id [](http://example.com/codeflow?start=11&length=2)

Note that this rule picks up some default information from the root location, such as the level of "note"

File: Src/Plugins/Security/SEC102.ReviewPotentiallySensitiveData.json [hunk 2]
```diff
@@ -22,6 +24,13 @@
...
+          "Name": "ReviewPotentiallySensitiveData/Url",
+          "FileNameAllowRegex": "(?i)(\\.cs|\\.cpp|\\.ini|\\.js|\\.ps1)$",
```
Comment (michaelcfanning) [inline Src/Plugins/Security/SEC102.ReviewPotentiallySensitiveData.json:31]: >FileNameAllowRegex [](http://example.com/codeflow?start=11&length=18)

This extensions list is a start to use when we run our investigative analysis. It's not clear this is what we'll want moving forward. Can discuss with you offline.

File: .gitignore [hunk 3]
```diff
@@ -226,4 +226,4 @@ ModelManifest.xml
 
 # SPAM configuration
 .spam/
-*.sarif
\ No newline at end of file
+.sarif/
```
Comment (michaelcfanning) [inline .gitignore:229]: >sarif [](http://example.com/codeflow?start=1&length=5)

The previous entry ignored newly added test files! Clearly not write. I think it was intended to ignore everything in the .sarif\ directory...

File: Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/ExpectedOutputs/SEC102_001.EmailAddress_with_email_addresses.sarif [hunk 4]
```diff
@@ -39,7 +39,7 @@
       "artifacts": [
         {
           "location": {
-            "uri": "src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/Inputs/SEC102_001.EmailAddress_with_email_addresses.txt",
```
Comment (michaelcfanning) [inline Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/ExpectedOutputs/SEC102_001.EmailAddress_with_email_addresses.sarif:42]: >EmailAddress_with_email_addresses [](http://example.com/codeflow?start=105&length=33)

Why did someone think suffixing this file with 'with_email_addresses' provided more information? I don't know. Seems redundant to me so I trimmed it. :)

File: Src/Plugins/Security/SEC102.ReviewPotentiallySensitiveData.json [hunk 5]
```diff
@@ -1,5 +1,7 @@
 {
   "ValidatorsAssemblyName": "Security.dll",
+  "SharedStringsFileName": "Security.SharedStrings.txt",
+
```
Comment (eddynaka) [inline Src/Plugins/Security/SEC102.ReviewPotentiallySensitiveData.json:4]: nit: remove this empty line

**Comment by eddynaka** [comment]
>             string candidateAddress = $"{id}.Value@{host?.Value}";

this looks strange.
I think it should be {id.Value}

---
Refers to: Src/Plugins/Security/SEC102_001.EmailAddressValidator.cs:22 in 35b0fc7. [](commit_id = 35b0fc7ab515f20d81906985d5aa225e6884706f, deletion_comment = False)

File: Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/ExpectedOutputs/SEC102_003.Url.sarif [hunk 6]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (eddynaka) [inline Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/ExpectedOutputs/SEC102_003.Url.sarif:1]: something is wrong. we should see results in the SARIF

File: Src/Plugins/Security/SEC102.ReviewPotentiallySensitiveData.json (hunk 2 above)
Comment (suvamM) [reply Src/Plugins/Security/SEC102.ReviewPotentiallySensitiveData.json:31]: @michaelcfanning Is this comment with respect to the URL analysis or the certificate analysis?

**Comment by suvamM** [comment]
@michaelcfanning @eddynaka I am trying to understand this: I see the test files in place (expected inputs and outputs), but I don't see any files which drive the tests to use this new test data. Will the existing testing setup pick these up automatically?

**Comment by michaelcfanning** [comment]
I'd neglected to git add a file, sorry!

---
In reply to: [1106972516](https://github.com/microsoft/sarif-pattern-matcher/pull/611#issuecomment-1106972516) [](http://example.com/codeflow?ancestors=1106972516,1106972516)

File: Src/Plugins/Security/SEC102.ReviewPotentiallySensitiveData.json (hunk 2 above)
Comment (michaelcfanning) [reply Src/Plugins/Security/SEC102.ReviewPotentiallySensitiveData.json:31]: URL

File: Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/ExpectedOutputs/SEC102_003.Url.sarif (hunk 6 above)
Comment (michaelcfanning) [reply Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/ExpectedOutputs/SEC102_003.Url.sarif:1]: Wrong extension on test file! Good catch, thank you.

**Comment by eddynaka** [comment]
> {

you can remove this entire file

---
Refers to: Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/ExpectedOutputs/SEC102_001.EmailAddress_with_email_addresses.sarif:1 in b9278ea. [](commit_id = b9278ead9073bcd10a45dd9b24b24f45b524aea0, deletion_comment = False)

**Comment by eddynaka** [comment]
> {

you can remove this entire file

---
Refers to: Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/ExpectedOutputs/SEC102_002.SocialSecurityNumber_with_social_security_numbers.sarif:1 in b9278ea. [](commit_id = b9278ead9073bcd10a45dd9b24b24f45b524aea0, deletion_comment = False)

File: Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/Inputs/SEC102_003.Url.ps1 [hunk 7]
```diff
@@ -0,0 +1,4 @@
+https://msn.com/test
+https://msn.com/test/
+https://msn.com/test?foo=bar
+https://msn.com/test?foo=bar#bookmark
```
Comment (eddynaka) [inline Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/Inputs/SEC102_003.Url.ps1:4]: In the tests, I cannot see this.

Comment (michaelcfanning) [reply Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/Inputs/SEC102_003.Url.ps1:4]: that's by design, this is a bookmark, i.e., the fragment doesn't change the finding.

**Comment by michaelcfanning** [comment]
> {

hm, the baselining gesture in my non-clean build env re-added them, thanks.

---
In reply to: [1109053602](https://github.com/microsoft/sarif-pattern-matcher/pull/611#issuecomment-1109053602) [](http://example.com/codeflow?ancestors=1109053602,1109053602)

---
Refers to: Src/Plugins/Tests.Security/TestData/ReviewPotentiallySensitiveData/ExpectedOutputs/SEC102_002.SocialSecurityNumber_with_social_security_numbers.sarif:1 in b9278ea. [](commit_id = b9278ead9073bcd10a45dd9b24b24f45b524aea0, deletion_comment = False)

**Review by eddynaka — APPROVED** [review]
:shipit:

**Pull request merged**